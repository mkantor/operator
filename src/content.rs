extern crate handlebars;
use crate::lib::*;
use handlebars::{Context, Handlebars, RenderContext, Renderable};
use serde::Serialize;
use std::io;
use std::path::{Path, PathBuf};
use thiserror::Error;

#[derive(Error, Debug)]
#[error(
  "Failed to parse template{} from content directory path '{}'.",
  .source.template_name.as_ref().map(|known_name| format!(" '{}'", known_name)).unwrap_or_default(),
  .content_directory_path.to_string_lossy()
)]
pub struct RegisteredTemplateParseError {
    source: handlebars::TemplateError,
    content_directory_path: PathBuf,
}

#[derive(Error, Debug)]
#[error(
  "Failed to parse template{}.",
  .source.template_name.as_ref().map(|known_name| format!(" '{}'", known_name)).unwrap_or_default(),
)]
pub struct UnregisteredTemplateParseError {
    source: handlebars::TemplateError,
}

#[derive(Error, Debug)]
pub enum ContentLoadingError {
    #[error(transparent)]
    TemplateParseError(#[from] RegisteredTemplateParseError),

    #[error(
    "Input/output error when loading{} from content directory path '{}'.",
    .name.as_ref().map(|known_name| format!(" '{}'", known_name)).unwrap_or(String::from(" content")),
    .content_directory_path.to_string_lossy()
  )]
    IOError {
        source: io::Error,
        content_directory_path: PathBuf,
        name: Option<String>,
    },
}

#[derive(Error, Debug)]
#[error(
  "Rendering failed for template{}.",
  .source.template_name.as_ref().map(|known_name| format!(" '{}'", known_name)).unwrap_or_default(),
)]
pub struct TemplateRenderError {
    #[from]
    source: handlebars::RenderError,
}

pub struct ContentEngine<'a> {
    template_registry: Handlebars<'a>,
}

impl<'a> ContentEngine<'a> {
    pub fn from_content_directory(
        content_directory_path: &'a Path,
    ) -> Result<Self, ContentLoadingError> {
        let template_registry = {
            let mut template_registry = Handlebars::new();
            template_registry
                .register_templates_directory(".hbs", content_directory_path)
                .map_err(|template_render_error| match template_render_error {
                    handlebars::TemplateFileError::TemplateError(source) => {
                        ContentLoadingError::TemplateParseError(RegisteredTemplateParseError {
                            source,
                            content_directory_path: PathBuf::from(content_directory_path),
                        })
                    }
                    handlebars::TemplateFileError::IOError(source, original_name) => {
                        // Handlebars-rust will use an empty string when the error does not
                        // correspond to a specific path.
                        let name = if original_name.is_empty() {
                            None
                        } else {
                            Some(original_name)
                        };
                        ContentLoadingError::IOError {
                            source,
                            content_directory_path: PathBuf::from(content_directory_path),
                            name,
                        }
                    }
                })?;
            template_registry
        };

        Ok(ContentEngine { template_registry })
    }

    pub fn new_content(
        &self,
        handlebars_source: &str,
    ) -> Result<ContentItem, UnregisteredTemplateParseError> {
        let template = handlebars::Template::compile2(handlebars_source, true)
            .map_err(|source| UnregisteredTemplateParseError { source })?;
        Ok(ContentItem::UnregisteredTemplate {
            template_registry: &self.template_registry,
            template,
        })
    }
}

#[derive(Serialize)]
struct GluonRenderData {
    version: GluonVersion,
}

#[derive(Serialize)]
struct RenderData {
    gluon: GluonRenderData,
}

pub enum ContentItem<'a> {
    /// A named template that exists in the registry. This variant must only be
    /// constructed with names that have already been validated against the
    /// registry.
    RegisteredTemplate {
        template_registry: &'a Handlebars<'a>,
        name_in_registry: &'a str,
    },

    /// An anonymous template for on-the-fly rendering.
    UnregisteredTemplate {
        template_registry: &'a Handlebars<'a>,
        template: handlebars::Template,
    },
}

impl Render for ContentItem<'_> {
    type RenderArgs = GluonVersion;
    type Error = TemplateRenderError;

    fn render(&self, gluon_version: GluonVersion) -> Result<String, Self::Error> {
        let render_data = RenderData {
            gluon: GluonRenderData {
                version: gluon_version,
            },
        };

        let rendered_content = match &self {
            ContentItem::RegisteredTemplate {
                template_registry,
                name_in_registry,
            } => template_registry.render(name_in_registry, &render_data)?,
            ContentItem::UnregisteredTemplate {
                template_registry,
                template,
            } => {
                let context = Context::wraps(render_data)?;
                let mut render_context = RenderContext::new(None);
                template.renders(template_registry, &context, &mut render_context)?
            }
        };

        Ok(rendered_content)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_lib::*;

    #[test]
    fn content_engine_can_be_created_from_valid_content_directory() {
        for &path in &CONTENT_DIRECTORY_PATHS_WITH_VALID_CONTENTS {
            assert!(
                ContentEngine::from_content_directory(Path::new(path)).is_ok(),
                "Content engine could not be created from {}",
                path
            );
        }
    }

    #[test]
    fn content_engine_cannot_be_created_from_invalid_content_directory() {
        for &path in &CONTENT_DIRECTORY_PATHS_WITH_INVALID_CONTENTS {
            assert!(
                ContentEngine::from_content_directory(Path::new(path)).is_err(),
                "Content engine was successfully created from {}, but this should have failed",
                path
            );
        }
    }

    #[test]
    fn new_templates_can_be_rendered() {
        let engine = ContentEngine::from_content_directory(
            arbitrary_content_directory_path_with_valid_content(),
        )
        .expect("Content engine could not be created");

        for &(template, expected_output) in &VALID_TEMPLATES {
            let new_content = engine
                .new_content(template)
                .expect("Template could not be parsed");
            let rendered = new_content
                .render(GluonVersion("0.0.0"))
                .expect(&format!("Template rendering failed for `{}`", template,));
            assert_eq!(
                rendered,
                expected_output,
                "Template rendering for `{}` did not produce the expected output (\"{}\"), instead got \"{}\"",
                template,
                expected_output,
                rendered,
            );
        }
    }

    #[test]
    fn new_content_fails_for_invalid_templates() {
        let engine = ContentEngine::from_content_directory(
            arbitrary_content_directory_path_with_valid_content(),
        )
        .expect("Content engine could not be created");

        for &template in &INVALID_TEMPLATES {
            let result = engine.new_content(template);

            assert!(
                result.is_err(),
                "Content was successfully created for invalid template `{}`, but it should have failed",
                template,
            );
        }
    }

    #[test]
    fn new_templates_can_reference_partials_from_content_directory() {
        let content_directory_path = example_path("valid/partials");
        let engine = ContentEngine::from_content_directory(&content_directory_path)
            .expect("Content engine could not be created");

        let template = "this is partial: {{> ab}}";
        let expected_output = "this is partial: a\nb\n\n";

        let new_content = engine
            .new_content(template)
            .expect("Template could not be parsed");
        let rendered = new_content
            .render(GluonVersion("0.0.0"))
            .expect(&format!("Template rendering failed for `{}`", template,));
        assert_eq!(
            rendered,
            expected_output,
            "Template rendering for `{}` did not produce the expected output (\"{}\"), instead got \"{}\"",
            template,
            expected_output,
            rendered,
        );
    }
}
