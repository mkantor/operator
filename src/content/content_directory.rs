use std::env;
use std::fs::File;
use std::os::unix::fs::PermissionsExt;
use std::path;
use std::path::{Path, PathBuf};
use thiserror::Error;
use walkdir::WalkDir;

#[derive(Error, Debug)]
pub enum ContentDirectoryFromRootError {
    #[error("Unable to use directory root '{}': {}", .root.display(), .message)]
    InvalidRootPath { root: PathBuf, message: String },

    #[error("Unable to use directory root '{}': {}", .root.display(), .source)]
    WalkDirError {
        root: PathBuf,
        source: walkdir::Error,
    },

    #[error(transparent)]
    DirectoryEntryError(#[from] ContentFileError),
}

#[derive(Error, Debug)]
#[error("Content file error: {}", .0)]
pub struct ContentFileError(String);

/// A filesystem directory containing content.
pub struct ContentDirectory {
    files: Vec<ContentFile>,
    #[cfg(test)]
    root: PathBuf,
}

impl ContentDirectory {
    pub fn from_root<P: AsRef<Path>>(
        absolute_root: &P,
    ) -> Result<Self, ContentDirectoryFromRootError> {
        let absolute_root_path = absolute_root.as_ref();
        if !absolute_root_path.is_absolute() {
            return Err(ContentDirectoryFromRootError::InvalidRootPath {
                message: String::from("Root path must be absolute."),
                root: PathBuf::from(absolute_root_path),
            });
        }

        let mut files = Vec::new();
        let walker = WalkDir::new(absolute_root_path)
            .follow_links(true)
            .min_depth(1);
        for dir_entry_result in walker {
            let dir_entry = dir_entry_result.map_err(|walkdir_error| {
                ContentDirectoryFromRootError::WalkDirError {
                    source: walkdir_error,
                    root: PathBuf::from(absolute_root_path),
                }
            })?;
            {
                let entry_path = dir_entry.path().to_path_buf();
                if dir_entry.file_type().is_file() {
                    let content_file =
                        ContentFile::from_root_and_path(absolute_root_path, entry_path)
                            .map_err(ContentDirectoryFromRootError::from)?;
                    files.push(content_file);
                }
            }
        }

        Ok(ContentDirectory {
            files,
            #[cfg(test)]
            root: absolute_root_path.into(),
        })
    }

    #[cfg(test)]
    pub fn root(&self) -> &Path {
        &self.root
    }
}

pub struct ContentFile {
    absolute_path: String,
    relative_path: String,
    is_hidden: bool,
    is_executable: bool,
    relative_path_without_extensions: String,
    extensions: Vec<String>,
    file: File,
}
impl ContentFile {
    pub const PATH_SEPARATOR: char = '/';

    fn from_root_and_path(
        content_directory_root: &Path,
        absolute_content_file_path: PathBuf,
    ) -> Result<Self, ContentFileError> {
        if path::MAIN_SEPARATOR != Self::PATH_SEPARATOR {
            return Err(ContentFileError(format!(
                "Platforms that use '{}' as a path separator are not supported",
                path::MAIN_SEPARATOR
            )));
        }

        let root = match content_directory_root.to_str() {
            Some(unicode_root) => unicode_root,
            None => {
                return Err(ContentFileError(format!(
                    "Non-unicode directory root (path is similar to '{}')",
                    content_directory_root.display(),
                )))
            }
        };

        let absolute_path = String::from(
            absolute_content_file_path
                .to_str()
                .ok_or_else(|| ContentFileError(String::from("Path was not unicode.")))?,
        );

        let relative_path = absolute_content_file_path
            .strip_prefix(root)
            .map_err(|strip_prefix_error| {
                ContentFileError(format!(
                    "Content file path '{}' did not start with expected prefix '{}': {}",
                    absolute_content_file_path.display(),
                    root,
                    strip_prefix_error
                ))
            })?
            .to_str()
            .map(String::from)
            .ok_or_else(|| ContentFileError(String::from("Path was not unicode.")))?;

        let file = File::open(&absolute_content_file_path).map_err(|io_error| {
            ContentFileError(format!(
                "Unable to open file '{}' in '{}' for reading: {}",
                relative_path, root, io_error
            ))
        })?;

        let basename = absolute_content_file_path
            .file_name()
            .ok_or_else(|| {
                ContentFileError(format!(
                    "Unable to get basename of '{}' in '{}'",
                    relative_path, root,
                ))
            })?
            .to_str()
            .ok_or_else(|| ContentFileError(String::from("File had a non-unicode basename.")))?;

        // Conventions around hidden files, whether a file is executable, etc
        // differ across platforms. It wouldn't be hard to implement this, but
        // Operator does not currently run its CI checks on non-unix platforms
        // so it would be too easy to introduce regressions.
        let (extensions, is_hidden, is_executable) = if !cfg!(unix) {
            return Err(ContentFileError(format!(
                "Operator does not currently support your operating system ({})",
                env::consts::OS,
            )));
        } else {
            // If the basename begins with `.` its first chunk isn't considered an "extension".
            let non_extension_components = if basename.starts_with('.') { 2 } else { 1 };
            let extensions = basename
                .split('.')
                .skip(non_extension_components)
                .map(String::from)
                .collect::<Vec<String>>();

            // The file is hidden if any of its relative path components starts
            // with a dot.
            let is_hidden = relative_path.starts_with('.')
                || relative_path.contains(&format!("{}.", Self::PATH_SEPARATOR));

            let permissions = file
                .metadata()
                .map_err(|io_error| {
                    ContentFileError(format!(
                        "Unable to query metadata for content file '{}': {}",
                        absolute_content_file_path.display(),
                        io_error
                    ))
                })?
                .permissions();
            let is_executable = permissions.mode() & 0o111 != 0;

            (extensions, is_hidden, is_executable)
        };

        let relative_path_without_extensions = String::from({
            let extensions_len = extensions.iter().fold(0, |len, extension| {
                // Extra 1 is to count . in the extensions.
                len + extension.len() + 1
            });
            &relative_path[0..(relative_path.len() - extensions_len)]
        });

        Ok(ContentFile {
            absolute_path,
            relative_path,
            relative_path_without_extensions,
            extensions,
            file,
            is_hidden,
            is_executable,
        })
    }

    pub fn absolute_path(&self) -> &str {
        &self.absolute_path
    }

    pub fn relative_path(&self) -> &str {
        &self.relative_path
    }

    pub fn is_hidden(&self) -> bool {
        self.is_hidden
    }

    pub fn is_executable(&self) -> bool {
        self.is_executable
    }

    pub fn relative_path_without_extensions(&self) -> &str {
        &self.relative_path_without_extensions
    }

    pub fn extensions(&self) -> &[String] {
        &self.extensions
    }

    pub fn file(&self) -> &File {
        &self.file
    }

    pub fn into_file(self) -> File {
        self.file
    }
}

impl IntoIterator for ContentDirectory {
    type Item = ContentFile;
    type IntoIter = std::vec::IntoIter<Self::Item>;
    fn into_iter(self) -> Self::IntoIter {
        self.files.into_iter()
    }
}

impl<'a> IntoIterator for &'a ContentDirectory {
    type Item = &'a ContentFile;
    type IntoIter = std::slice::Iter<'a, ContentFile>;
    fn into_iter(self) -> Self::IntoIter {
        self.files.iter()
    }
}

#[cfg(test)]
impl PartialEq for ContentDirectory {
    fn eq(&self, other: &Self) -> bool {
        self.root() == other.root()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_lib::*;
    use std::fs;

    #[test]
    fn directory_can_be_created_from_valid_root() {
        let path = fs::canonicalize("./src").expect("Canonicalizing path failed");
        let result = ContentDirectory::from_root(&path);
        assert!(
            result.is_ok(),
            "Unable to use directory at '{}': {}",
            path.display(),
            result.err().unwrap().to_string()
        );
    }

    #[test]
    fn directory_root_must_exist() {
        let result = ContentDirectory::from_root(&example_path("this/does/not/actually/exist"));
        assert!(
            result.is_err(),
            "Directory was successfully created from non-existent path",
        );
    }

    #[test]
    fn directory_root_must_be_absolute_path() {
        let non_absolute_path = "./src";
        let result = ContentDirectory::from_root(&non_absolute_path);
        assert!(
            result.is_err(),
            "ContentDirectory was successfully created from non-absolute path '{}', but this should have failed",
            non_absolute_path,
        );
    }
}
