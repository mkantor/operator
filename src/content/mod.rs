mod body;
mod content_directory;
mod content_engine;
mod content_index;
mod content_item;
mod content_registry;
mod handlebars_helpers;
mod mime;
mod route;
mod test_lib;

use crate::bug_message;
use bytes::Bytes;
use content_item::RenderingFailedError;
use futures::Stream;
use serde::Serialize;
use std::io;
use thiserror::Error;

pub use self::mime::{MediaRange, MediaType};
pub use content_directory::ContentDirectory;
pub use content_engine::{
    ContentEngine, ContentLoadingError, FilesystemBasedContentEngine, TemplateError,
};
pub use content_index::ContentIndex;
pub use content_item::UnregisteredTemplate;
pub use content_registry::{ContentRepresentations, RegisteredContent};
pub use route::Route;

// This is just a trait alias to help make type signatures a bit saner.
pub trait ByteStream: Stream<Item = Result<Bytes, StreamError>>
where
    Self: Unpin,
{
}
impl<T> ByteStream for T where T: Stream<Item = Result<Bytes, StreamError>> + Unpin {}

/// A piece of rendered content along with its media type.
pub struct Media<Content: ByteStream> {
    pub media_type: MediaType,
    pub content: Content,
}
impl<Content: ByteStream> Media<Content> {
    fn new(media_type: MediaType, content: Content) -> Self {
        Self {
            media_type,
            content,
        }
    }
}

/// Indicates that it was not possible to produce rendered output, either
/// because rendering was attempted and failed or because no acceptable media
/// types are available.
#[derive(Error, Debug)]
pub enum RenderError {
    #[error(transparent)]
    RenderingFailed(RenderingFailedError),

    #[error("The requested content cannot be rendered as an acceptable media type.")]
    CannotProvideAcceptableMediaType,

    #[doc(hidden)]
    #[error("{} This should never happen: {}", bug_message!(), .0)]
    Bug(String),
}

/// Indicates that something went wrong after starting to stream content.
#[derive(Error, Debug)]
pub enum StreamError {
    #[error(
        "Process exited with {}{}",
        match .exit_code {
            Some(code) => format!("code {}", code),
            None => String::from("unknown code"),
        },
        .stderr_contents.as_ref().map(|message| format!(": {}", message)).unwrap_or_default(),
    )]
    ExecutableExitedWithNonzero {
        pid: u32,
        exit_code: Option<i32>,
        stderr_contents: Option<String>,
    },

    #[error("Executable output could not be captured")]
    ExecutableOutputCouldNotBeCaptured { pid: u32 },

    #[error("Input/output error during rendering")]
    IOError {
        #[from]
        source: io::Error,
    },

    #[error("Stream was cancelled")]
    Canceled,
}

pub trait Render {
    type Output;
    fn render<'engine, 'accept, ServerInfo, QueryParameters, Engine, Accept>(
        &self,
        context: RenderContext<'engine, ServerInfo, QueryParameters, Engine>,
        acceptable_media_ranges: Accept,
    ) -> Result<Media<Self::Output>, RenderError>
    where
        ServerInfo: Clone + Serialize,
        QueryParameters: Clone + Serialize,
        Engine: ContentEngine<ServerInfo>,
        Accept: IntoIterator<Item = &'accept MediaRange>,
        Self::Output: ByteStream;
}

// These must match up with serialized property names in RequestData and
// RenderData.
const TARGET_MEDIA_TYPE_PROPERTY_NAME: &str = "target-media-type";
const REQUEST_DATA_PROPERTY_NAME: &str = "request";
const ROUTE_PROPERTY_NAME: &str = "route";
const QUERY_PARAMETERS_PROPERTY_NAME: &str = "query-parameters";

/// Render data that comes from requests.
#[derive(Clone, Serialize)]
#[serde(rename_all = "kebab-case")]
pub struct RequestData<QueryParameters: Clone + Serialize> {
    /// The request [`Route`] that caused this content to be rendered, if any.
    pub route: Option<Route>,

    /// A parsed version of the request URI's query string.
    pub query_parameters: QueryParameters,
}

/// Data passed to handlebars templates and executables.
///
/// Fields serialize into kebab-case (e.g. `server_info` becomes `server-info`).
#[derive(Clone, Serialize)]
#[serde(rename_all = "kebab-case")]
pub struct RenderData<ServerInfo: Clone + Serialize, QueryParameters: Clone + Serialize> {
    /// A hierarchial index of the content. This is serialized with the name
    /// `/` (with handlebars escaping this looks like `[/].[foo/].bar`).
    #[serde(rename = "/")]
    pub index: ContentIndex,

    /// Metadata about the server, such as its version.
    pub server_info: ServerInfo,

    /// The best [`MediaType`] as determined by content negotiation. Rendering
    /// must emit content in this media type.
    pub target_media_type: Option<MediaType>,

    /// Data that comes from requests.
    pub request: RequestData<QueryParameters>,

    /// An [HTTP `4xx` or `5xx` status code](https://datatracker.ietf.org/doc/html/rfc7231#section-6)
    /// indicating that something went wrong. This will be set while rendering
    /// content for the `--error-handler-route`.
    pub error_code: Option<u16>,
}

/// Values used during rendering, including the data passed to handlebars
/// templates and executables.
pub struct RenderContext<'engine, ServerInfo, QueryParameters, Engine>
where
    ServerInfo: Clone + Serialize,
    QueryParameters: Clone + Serialize,
    Engine: ContentEngine<ServerInfo>,
{
    content_engine: &'engine Engine,
    data: RenderData<ServerInfo, QueryParameters>,
}

impl<'engine, ServerInfo, QueryParameters, Engine>
    RenderContext<'engine, ServerInfo, QueryParameters, Engine>
where
    ServerInfo: Clone + Serialize,
    Engine: ContentEngine<ServerInfo>,
    QueryParameters: Clone + Serialize,
{
    pub fn into_error_context(self, error_code: u16) -> Self {
        RenderContext {
            data: RenderData {
                error_code: Some(error_code),
                ..self.data
            },
            ..self
        }
    }
}
