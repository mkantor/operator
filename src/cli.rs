use crate::content::*;
use crate::content_directory::ContentDirectory;
use crate::http;
use crate::lib::*;
use std::io;
use std::net::ToSocketAddrs;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum RenderCommandError {
    #[error("Failed to read input.")]
    ReadError { source: io::Error },

    #[error("Unable to load content.")]
    ContentLoadingError {
        #[from]
        source: ContentLoadingError,
    },

    #[error("Unable to parse template from content directory.")]
    RegisteredTemplateParseError {
        #[from]
        source: RegisteredTemplateParseError,
    },

    #[error("Unable to parse template from input.")]
    UnregisteredTemplateParseError {
        #[from]
        source: UnregisteredTemplateParseError,
    },

    #[error("Unable to render content.")]
    ContentRenderingError {
        #[from]
        source: ContentRenderingError,
    },

    #[error("Failed to write output.")]
    WriteError { source: io::Error },
}

#[derive(Error, Debug)]
pub enum GetCommandError {
    #[error("Unable to load content.")]
    ContentLoadingError {
        #[from]
        source: ContentLoadingError,
    },

    #[error("Unable to parse template from content directory.")]
    RegisteredTemplateParseError {
        #[from]
        source: RegisteredTemplateParseError,
    },

    #[error("Content not found at route '{}'.", .route)]
    ContentNotFound { route: String },

    #[error("Unable to render content.")]
    ContentRenderingError {
        #[from]
        source: ContentRenderingError,
    },

    #[error("Failed to write output.")]
    WriteError { source: io::Error },
}

#[derive(Error, Debug)]
pub enum ServeCommandError {
    #[error("Unable to load content.")]
    ContentLoadingError {
        #[from]
        source: ContentLoadingError,
    },

    #[error("Failed to run server.")]
    ServerError { source: io::Error },
}

/// Reads a template from `input`, renders it, and writes it to `output`.
pub fn render<I: io::Read, O: io::Write>(
    content_directory: ContentDirectory,
    media_type: MediaType,
    soliton_version: SolitonVersion,
    input: &mut I,
    output: &mut O,
) -> Result<(), RenderCommandError> {
    let shared_content_engine =
        FilesystemBasedContentEngine::from_content_directory(content_directory, soliton_version)?;
    let content_engine = shared_content_engine
        .read()
        .expect("RwLock for ContentEngine has been poisoned");

    let mut template = String::new();
    input
        .read_to_string(&mut template)
        .map_err(|source| RenderCommandError::ReadError { source })?;

    let content_item = content_engine.new_template(&template, media_type.clone())?;
    let render_context = content_engine.get_render_context();
    let rendered_output = content_item.render(render_context, &[media_type.as_media_range()])?;
    write!(output, "{}", rendered_output)
        .map_err(|source| RenderCommandError::WriteError { source })?;

    output
        .flush()
        .map_err(|source| RenderCommandError::WriteError { source })
}

/// Renders an item from the content directory and writes it to `output`.
pub fn get<O: io::Write>(
    content_directory: ContentDirectory,
    route: &str,
    accept: MediaRange,
    soliton_version: SolitonVersion,
    output: &mut O,
) -> Result<(), GetCommandError> {
    let shared_content_engine =
        FilesystemBasedContentEngine::from_content_directory(content_directory, soliton_version)?;
    let content_engine = shared_content_engine
        .read()
        .expect("RwLock for ContentEngine has been poisoned");

    let content_item = content_engine
        .get(route)
        .ok_or(GetCommandError::ContentNotFound {
            route: String::from(route),
        })?;
    let render_context = content_engine.get_render_context();
    let rendered_output = content_item.render(render_context, &[accept])?;
    write!(output, "{}", rendered_output)
        .map_err(|source| GetCommandError::WriteError { source })?;

    output
        .flush()
        .map_err(|source| GetCommandError::WriteError { source })
}

/// Starts an HTTP server for the given content directory.
pub fn serve<A: 'static + ToSocketAddrs>(
    content_directory: ContentDirectory,
    index_route: &str,
    socket_address: A,
    soliton_version: SolitonVersion,
) -> Result<(), ServeCommandError> {
    let shared_content_engine =
        FilesystemBasedContentEngine::from_content_directory(content_directory, soliton_version)?;

    http::run_server(shared_content_engine, index_route, socket_address)
        .map_err(|source| ServeCommandError::ServerError { source })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_lib::*;
    use mime_guess::MimeGuess;
    use std::collections::{BTreeMap, HashMap};
    use std::str;

    /// Attempts to render all non-hidden files in ContentDirectory, returning
    /// them as a map of Route -> RenderedContent | ErrorMessage.
    fn render_everything(
        content_directory: ContentDirectory,
    ) -> Result<HashMap<String, String>, String> {
        let mut content = HashMap::new();
        let content_directory_root = content_directory.root();
        for content_file in content_directory {
            if !content_file.is_hidden() {
                // Create a separate ContentDirectory that can be consumed by
                // the CLI calls.
                let consumable_content_directory =
                    ContentDirectory::from_root(&content_directory_root).map_err(|error| {
                        format!("Could not create content directory: {:?}", error)
                    })?;
                let route = content_file.relative_path_without_extensions();
                let empty_string = String::from("");
                let first_filename_extension =
                    content_file.extensions().first().unwrap_or(&empty_string);

                // Target media type is just the source media type. This isn't
                // testing transcoding.
                let target_media_type = MimeGuess::from_ext(first_filename_extension)
                    .first()
                    .unwrap_or(mime::APPLICATION_OCTET_STREAM);

                let mut output = Vec::new();
                let result = get(
                    consumable_content_directory,
                    route,
                    target_media_type,
                    SolitonVersion("0.0.0"),
                    &mut output,
                );

                let output_or_error_message = match result {
                    Ok(()) => String::from_utf8(output)
                        .map_err(|error| format!("Output was not valid UTF-8: {:?}", error))?,
                    Err(error) => {
                        let anyhow_error = anyhow::Error::from(error);
                        let causes = anyhow_error.chain().map(|error| error.to_string());
                        let message = causes.fold(String::new(), |acc, arg| acc + " " + &arg);
                        format!("Error:{}", message)
                    }
                };

                content.insert(
                    String::from(content_file.relative_path_without_extensions()),
                    output_or_error_message,
                );
            }
        }
        Ok(content)
    }

    #[test]
    fn examples_match_snapshots() {
        for content_directory in example_content_directories() {
            let content_directory_root = &content_directory.root();

            let unordered_content =
                render_everything(content_directory).expect("Fatal error in valid example");
            let contents = unordered_content
                .iter()
                // If dynamic content files mention where the repo is checked
                // out, redact it to keep tests portable.
                .map(|(key, value)| (key, value.replace(PROJECT_DIRECTORY, "$PROJECT_DIRECTORY")))
                // Files can be omitted from snapshots with a naming convention.
                .filter(|(key, _)| !key.ends_with("-SKIP-SNAPSHOT"))
                .collect::<BTreeMap<_, _>>();

            let mut insta_settings = insta::Settings::clone_current();
            insta_settings.set_input_file(content_directory_root);
            let id = content_directory_root
                .strip_prefix(example_path("."))
                .unwrap()
                .to_string_lossy();
            insta_settings.set_snapshot_suffix(id);
            insta_settings.bind(|| insta::assert_yaml_snapshot!(contents));
        }
    }

    #[test]
    fn valid_templates_can_be_rendered() {
        for &(template, expected_output) in &VALID_TEMPLATES {
            let mut input = template.as_bytes();
            let mut output = Vec::new();
            let directory = arbitrary_content_directory_with_valid_content();
            let result = render(
                directory,
                MediaType::from_media_range(mime::TEXT_HTML).unwrap(),
                SolitonVersion("0.0.0"),
                &mut input,
                &mut output,
            );

            assert!(
                result.is_ok(),
                "Template rendering failed for `{}`: {}",
                template,
                result.unwrap_err(),
            );
            let output_as_str = str::from_utf8(output.as_slice()).expect("Output was not UTF-8");
            assert_eq!(
                output_as_str,
                expected_output,
                "Template rendering for `{}` did not produce the expected output (\"{}\"), instead got \"{}\"",
                template,
                expected_output,
                output_as_str
            );
        }
    }

    #[test]
    fn invalid_templates_cannot_be_rendered() {
        for &template in &INVALID_TEMPLATES {
            let mut input = template.as_bytes();
            let mut output = Vec::new();
            let directory = arbitrary_content_directory_with_valid_content();
            let result = render(
                directory,
                MediaType::from_media_range(mime::TEXT_HTML).unwrap(),
                SolitonVersion("0.0.0"),
                &mut input,
                &mut output,
            );

            assert!(
                result.is_err(),
                "Template rendering succeeded for `{}`, but it should have failed",
                template,
            );
        }
    }

    #[test]
    fn content_can_be_retrieved_from_content_directory() {
        let mut output = Vec::new();
        let route = "hello";
        let expected_output = "hello world\n";

        let directory = arbitrary_content_directory_with_valid_content();
        let result = get(
            directory,
            route,
            mime::TEXT_HTML,
            SolitonVersion("0.0.0"),
            &mut output,
        );

        assert!(
            result.is_ok(),
            "Template rendering failed for content at '{}': {}",
            route,
            result.unwrap_err(),
        );
        let output_as_str = str::from_utf8(output.as_slice()).expect("Output was not UTF-8");
        assert_eq!(
            output_as_str,
            expected_output,
            "Template rendering for content at '{}' did not produce the expected output (\"{}\"), instead got \"{}\"",
            route,
            expected_output,
            output_as_str
        );
    }

    #[test]
    fn getting_content_which_does_not_exist_is_an_error() {
        let mut output = Vec::new();
        let route = "this-route-does-not-refer-to-any-content";

        let directory = arbitrary_content_directory_with_valid_content();
        let result = get(
            directory,
            route,
            mime::TEXT_HTML,
            SolitonVersion("0.0.0"),
            &mut output,
        );

        match result {
            Ok(_) => panic!(
                "Getting content from '{}' succeeded, but it should have failed",
                route
            ),
            Err(GetCommandError::ContentNotFound {
                route: route_from_error,
            }) => assert_eq!(
                route_from_error, route,
                "Route from error did not match route used"
            ),
            Err(_) => panic!("Wrong type of error was produced, expected ContentNotFound"),
        };
    }
}
