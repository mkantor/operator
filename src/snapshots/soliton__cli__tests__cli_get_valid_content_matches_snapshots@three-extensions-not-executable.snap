---
source: src/cli.rs
expression: contents
input_file: examples/invalid/three-extensions-not-executable
---
this-file-is-not-executable: "Error: Unable to load content. Content file name is not supported: Content file name 'this-file-is-not-executable.txt.html.jpg' has too many extensions. "
